src/artifactory_path.rs

```diff
@@ -0,0 +1,150 @@
+static VALID_OS: &[&str] = &["windows", "macos", "linux"];
+static VALID_ARCH: &[&str] = &["x86_64", "arm64", "aarch64", "i686"];
+
+pub fn generate_artifactory_path<S: Into<String>>(
+    repo: S,
+    tool_name: S,
+    version: S,
+    operating_system: S,
+    architecture: Option<S>,
+) -> ForemanResult<String> {
+    let repo = repo.into();
+    let tool_name = tool_name.into();
+    let version = version.into();
+    let operating_system = operating_system.into();
+
+    check_valid_os(&operating_system)?;
+    check_valid_version(&version)?;
+    let mut full_tool_name = format!("{}-{}-{}", tool_name, version, operating_system);
+    if let Some(architecture) = architecture {
+        let architecture = architecture.into();
+        check_valid_arch(&architecture)?;
+        full_tool_name.push('-');
+        full_tool_name.push_str(&architecture);
+    }
+
+    full_tool_name.push_str(".zip");
+
+    Ok(format!(
+        "artifactory/{}/{}/{}/{}",
+        repo, tool_name, version, full_tool_name
+    ))
+}
+
+fn check_valid_os(operating_system: &str) -> ForemanResult<()> {
+    if !VALID_OS.contains(&operating_system) {
+        return Err(ForemanError::io_error_with_context(
+            Error::new(ErrorKind::InvalidInput, "Invalid Argument"),
+            format!(
+                "Invalid operating system: {}. Please input a valid operating system: {}",
+                operating_system,
+                VALID_OS.join(", ")
+            ),
+        ));
+    } else {
+        Ok(())
+    }
+}
+
+fn check_valid_arch(architecture: &str) -> ForemanResult<()> {
+    if !VALID_ARCH.contains(&architecture) {
+        return Err(ForemanError::io_error_with_context(
+            Error::new(ErrorKind::InvalidInput, "Invalid Argument"),
+            format!(
+                "Invalid architecture: {}. Please input a valid architecture: {}",
+                architecture,
+                VALID_ARCH.join(", ")
+            ),
+        ));
+    } else {
+        Ok(())
+    }
+}
+
+fn check_valid_version(version: &str) -> ForemanResult<()> {
+    if !version.starts_with('v') {
+        return Err(ForemanError::io_error_with_context(
+            Error::new(ErrorKind::InvalidInput, "Invalid Argument"),
+            format!("Invalid version: {}. Versions must start with a v", version),
+        ));
+    }
+
+    if let Err(err) = Version::parse(&version[1..]) {
+        Err(ForemanError::io_error_with_context(
+            Error::new(ErrorKind::InvalidInput, "Invalid Argument"),
+            format!("Invalid version: {}. Error: {}", version, err),
+        ))
+    } else {
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::generate_artifactory_path;
+
+    #[test]
+    fn simple_path() {
+        let path = generate_artifactory_path("repo", "tool_name", "v0.1.0", "macos", None).unwrap();
+        assert_eq!(
+            path,
+            "artifactory/repo/tool_name/v0.1.0/tool_name-v0.1.0-macos.zip"
```

Comment:
One interesting detail here is that the version number isn't duplicated in the github releases, but it also probably isn't a problem (we only really look for the os/arch keywords, iirc). I think we should be fine as is, but it's something to call attention to in case it affects our ability to share the artifact-choosing/download implementation.